kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -127,74 +160,134 @@ pub fn create<F: PrimeField>(
 
     let foreign_modulus = ForeignElement::from_biguint(modulus);
 
-    // Create multi-range-check witness for first left input
     let mut left = ForeignElement::from_biguint(inputs[LO].clone());
-    range_check::witness::extend(&mut witness, left.clone());
-    let mut add_values: Vec<(F, F, F, F)> = vec![];
+    // Create multi-range-check witness for first left input
+    if range_checks {
+        range_check::witness::extend(&mut witness, left.clone());
+    }
+    let mut values: Vec<FFAddValues<F>> = vec![];
     for i in 0..num {
         let right = ForeignElement::from_biguint(inputs[i + 1].clone());
-        let (output, sign, overflow, carry_lo, carry_mi) =
-            compute_subadd_values(&left, &right, opcodes[i], &foreign_modulus);
-        // Create multi-range-check witness for right_input (left_input was done in previous iteration) and output
+        let add_values = compute_ffadd_values(&left, &right, opcodes[i], &foreign_modulus);
         // We only obtain the 3 lower limbs of right because the range check takes only 264 bits now
         let right_3_limb = ForeignElement::new([right[LO], right[MI], right[HI]]);
-        range_check::witness::extend(&mut witness, right_3_limb);
-        range_check::witness::extend(&mut witness, output.clone());
+        // Create multi-range-check witness for right_input (left_input was done in previous iteration) and output
+        if range_checks {
+            range_check::witness::extend(&mut witness, right_3_limb);
```

Comment:
Once updated against master, you can just do `range_check::witness::extend(&mut witness, right.limbs);` and avoid the extra code of `let right_3_limb = ForeignElement::new([right[LO], right[MI], right[HI]]);`